### ISSUE TITLE
PyLance does not properly resolve TensorFlow modules

### ISSUE THREAD

**Issue by BlueskyFR** [microsoft/pylance-release#1066]
<h3> Environment data </h3>

-   Language Server version: 2021.3.2
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: basic

<h3> Problem </h3>

PyLance does not provide proper autocompletion for TensorFlow modules such as `tensorflow.data`:
![image](https://user-images.githubusercontent.com/25421460/111600992-39f26180-87d2-11eb-8be3-e8389f2a64c1.png)

Instead, we have to go through the `tensorflow.python` namespace which is not the recommended way to use it:
![image](https://user-images.githubusercontent.com/25421460/111601236-732ad180-87d2-11eb-8a13-750ec39e5f19.png)

<h3> Code Snippet / Additional information </h3>

**Works but is not the recommended way to import the module**:
```python
from tensorflow.python.data import Dataset
```

**Works but PyLance wrongly reports an error**:
```python
from tensorflow.data import Dataset
```
![image](https://user-images.githubusercontent.com/25421460/111601513-b7b66d00-87d2-11eb-9c43-38a064a3f7ab.png)
Same but in Python console:
![image](https://user-images.githubusercontent.com/25421460/111601588-cdc42d80-87d2-11eb-90f6-47354f4b004a.png)

I believe that this behavior is not exclusive to the TensorFlow library but for many other ones too.

Please feel free to ask if you need any more details!

**Comment by judej** [comment]
Thanks for the report. There are several challenges in providing type info for TensoFlow. We will look at ways to address this and get type info in Pylance.

**Comment by bschnurr** [comment]
tensorflow exposes the data api's using some dynamic code which pylance doesn't know how to parse.

Do any of these suggestions help you?

`from tensorflow._api.v2.data import Dataset`

or 

```
import tensorflow as tf
tf.data.Dataset
```

**Comment by bschnurr** [comment]
Now after I see how data is being imported I think this is a valid bug.

**Comment by bschnurr** [comment]
Ok. this test code runs 

```
from tensorflow.data import Dataset

dataset = Dataset.from_tensor_slices([1, 2, 3])
for element in dataset:
  print(element)
```

but pylance gives and error due to tensorflow dynamically modifying sys path in tensorflow/__ init __.py


```
# Make sure directory containing top level submodules is in
# the __path__ so that "from tensorflow.foo import bar" works.
# We're using bitwise, but there's nothing special about that.
_API_MODULE = _sys.modules[__name__].bitwise
_tf_api_dir = _os.path.dirname(_os.path.dirname(_API_MODULE.__file__))
_current_module = _sys.modules[__name__]

if not hasattr(_current_module, '__path__'):
  __path__ = [_tf_api_dir]
elif _tf_api_dir not in __path__:
  __path__.append(_tf_api_dir)
```

**Comment by hmc-cs-mdrissi** [comment]
Another major module that seems to have import issues is tensorflow.keras using pyright generated stubs. The stubs are necessary for me to encounter this issue.
<img width="966" alt="Screen Shot 2021-03-25 at 1 47 49 AM" src="https://user-images.githubusercontent.com/16809055/112430487-3ac44f80-8d0c-11eb-9790-37e3a4a817a4.png">

The weird thing for me is when I used pyright --createstub there is a tensorflow/keras folder in typing, but it still struggles to detect that module. At the same time weirdly auto-complete is happy when I try to type import,

<img width="848" alt="Screen Shot 2021-03-25 at 1 54 36 AM" src="https://user-images.githubusercontent.com/16809055/112431157-259bf080-8d0d-11eb-9ef6-55328096c5f7.png">

If I remove the pyright stubs then it stops complaining about keras. Python version is 3.7 and tensorflow version is 2.2. You should be able to reproduce it by trying,

```
import tensorflow as tf

layer = tf.keras.layers.Dense(5)
```

pyright on that script, see no errors, then do pyright --createstub tensoorflow, and run pyright again on that same script and it justs complaining about keras.

**Comment by bschnurr** [comment]
this issue with missing tensorflow submodules also seems similar to this pygame issue. https://github.com/microsoft/pylance-release/issues/758#issuecomment-806997580

**Comment by bschnurr** [comment]
> Another major module that seems to have import issues is tensorflow.keras using pyright generated stubs. The stubs are necessary for me to encounter this issue.
> <img alt="Screen Shot 2021-03-25 at 1 47 49 AM" width="966" src="https://user-images.githubusercontent.com/16809055/112430487-3ac44f80-8d0c-11eb-9790-37e3a4a817a4.png">
> 
> The weird thing for me is when I used pyright --createstub there is a tensorflow/keras folder in typing, but it still struggles to detect that module. At the same time weirdly auto-complete is happy when I try to type import,
> 
> <img alt="Screen Shot 2021-03-25 at 1 54 36 AM" width="848" src="https://user-images.githubusercontent.com/16809055/112431157-259bf080-8d0d-11eb-9ef6-55328096c5f7.png">
> 
> If I remove the pyright stubs then it stops complaining about keras. Python version is 3.7 and tensorflow version is 2.2. You should be able to reproduce it by trying,
> 
> ```
> import tensorflow as tf
> 
> layer = tf.keras.layers.Dense(5)
> ```
> 
> pyright on that script, see no errors, then do pyright --createstub tensoorflow, and run pyright again on that same script and it justs complaining about keras.

I believe this is due to the fact that pyright/pylance rely on stub folder structures matching. But the original source doesn't have a tensorflow/keras folder but is making it appear like it does.  Maybe try mirroring the keras stub folder to where it is in source.

**Comment by hmc-cs-mdrissi** [comment]
Curious. I'll make an issue later trying to better reproduce this (using a mac), but trying tensorflow 2.4.1 on a new virtualenv I still see the same issue of keras not a known member. I'll see if I can make a simple docker image that shows the problem.

**Comment by jakebailey** [comment]
Retesting this with the latest version of tensorflow; this doesn't work:

![image](https://user-images.githubusercontent.com/5341706/134093557-e9ddea4b-3a17-4fc7-b378-9f836df9e40f.png)

Opening `tensorflow/__init__.py` shows: 
![image](https://user-images.githubusercontent.com/5341706/134093575-34ece9bf-b673-4e07-b946-2cdfed26deca.png)

But, one can do:

![image](https://user-images.githubusercontent.com/5341706/134093626-3ca94aba-c8af-4f9d-8d57-ebedb20523d5.png)

I can't tell if this is a bug (and we should be showing `data` as a valid import because it's a module) or not. I can see this used a handful of times across GitHub, so it must be able to execute (or, maybe did with TFv1).

Regarding `tf.keras`, that exists, but is a special class that enables lazy loading, but completely obscures the type of the module:

![image](https://user-images.githubusercontent.com/5341706/134093708-be742976-38bf-4a70-8002-55b2eb3f5baf.png)

**Comment by bschnurr** [comment]
try adding this to the bottom of your  `tensorflow/__init__.py`
```
# Explicitly import lazy-loaded modules to support autocompletion.
# pylint: disable=g-import-not-at-top
if _typing.TYPE_CHECKING:
  from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
  from keras.api._v2 import keras
  from keras.api._v2.keras import losses
  from keras.api._v2.keras import metrics
  from keras.api._v2.keras import optimizers
  from keras.api._v2.keras import initializers
# pylint: enable=g-import-not-at-top
```

**Comment by nyngwang** [comment]
> try adding this to the bottom of your `tensorflow/__init__.py`
> 
> ```
> # Explicitly import lazy-loaded modules to support autocompletion.
> # pylint: disable=g-import-not-at-top
> if _typing.TYPE_CHECKING:
>   from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
>   from keras.api._v2 import keras
>   from keras.api._v2.keras import losses
>   from keras.api._v2.keras import metrics
>   from keras.api._v2.keras import optimizers
>   from keras.api._v2.keras import initializers
> # pylint: enable=g-import-not-at-top
> ```

Hihi, any progress here? Is there any option to disable lazy loading to enable auto-completion without these kind of hack?

**Comment by fang6100146** [comment]
> try adding this to the bottom of your `tensorflow/__init__.py`
> 
> ```
> # Explicitly import lazy-loaded modules to support autocompletion.
> # pylint: disable=g-import-not-at-top
> if _typing.TYPE_CHECKING:
>   from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
>   from keras.api._v2 import keras
>   from keras.api._v2.keras import losses
>   from keras.api._v2.keras import metrics
>   from keras.api._v2.keras import optimizers
>   from keras.api._v2.keras import initializers
> # pylint: enable=g-import-not-at-top
> ```
Thanks! It worked on my machine

**Comment by bschnurr** [comment]
> > try adding this to the bottom of your `tensorflow/__init__.py`
> > ```
> > # Explicitly import lazy-loaded modules to support autocompletion.
> > # pylint: disable=g-import-not-at-top
> > if _typing.TYPE_CHECKING:
> >   from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
> >   from keras.api._v2 import keras
> >   from keras.api._v2.keras import losses
> >   from keras.api._v2.keras import metrics
> >   from keras.api._v2.keras import optimizers
> >   from keras.api._v2.keras import initializers
> > # pylint: enable=g-import-not-at-top
> > ```
> 
> Hihi, any progress here? Is there any option to disable lazy loading to enable auto-completion without this kind of hack?

waiting on tensorflow
https://github.com/tensorflow/tensorflow/pull/54104

**Comment by nyngwang** [comment]
To @hmc-cs-mdrissi:

> [...]
> The weird thing for me is when I used pyright [...]

FYI: I found pylsp provides much better UX in linting(i.e. auto-formatting on-save) and pyright seems to create a(or two) `node` process(es) that occupy 99% of CPU usage. I will never trust Micro$oft again.

**Comment by imanbabaei** [comment]
> try adding this to the bottom of your `tensorflow/__init__.py`
> 
> ```
> # Explicitly import lazy-loaded modules to support autocompletion.
> # pylint: disable=g-import-not-at-top
> if _typing.TYPE_CHECKING:
>   from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
>   from keras.api._v2 import keras
>   from keras.api._v2.keras import losses
>   from keras.api._v2.keras import metrics
>   from keras.api._v2.keras import optimizers
>   from keras.api._v2.keras import initializers
> # pylint: enable=g-import-not-at-top
> ```

Thank you! It worked. Same problem in Pycharm. python 3.8.10, tensorflow 2.8

**Comment by bschnurr** [comment]
tf pr has been merged into main. 
https://github.com/tensorflow/tensorflow/pull/54104#event-6398616699

**Comment by heejaechang** [comment]
looks like it is already fixed by upstream through @bschnurr 's PR.

**Comment by bschnurr** [comment]
I dont see my tensorflow changes in 2.9.1 yet.

**Comment by YouJiacheng** [comment]
Is it proper to close this issue?
@bschnurr 's PR only fix `import tensorflow as tf; tf.keras` but not `from tensorflow.keras import ...`

**Comment by rezaneam** [comment]
@bschnurr 
Got it. Your changes were merged on Apr 8th, so I thought it should be part of 2.9.1. Honestly, didn't check when 2.9.1 was branched.
Your suggestion of using `tf.keras.preprocessing` instead sounds fair.

**Comment by leaver2000** [comment]
a pretty straight forward workaround
``` python
from importlib import import_module

import tensorflow as tf
import keras
from keras.api._v2 import keras as KerasAPI
# using the import module import the tensorflow.keras module
# and typehint that the type is KerasAPI module
keras: KerasAPI = import_module("tensorflow.keras")
```
![image](https://user-images.githubusercontent.com/76945789/182495025-d8f61702-51ab-4dda-bf6f-28c9f46b0730.png)

**Comment by Luffffffy** [comment]
> 尝试将此添加到您的底部 `tensorflow/__init__.py`
> 
> ```
> # Explicitly import lazy-loaded modules to support autocompletion.
> # pylint: disable=g-import-not-at-top
> if _typing.TYPE_CHECKING:
>   from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
>   from keras.api._v2 import keras
>   from keras.api._v2.keras import losses
>   from keras.api._v2.keras import metrics
>   from keras.api._v2.keras import optimizers
>   from keras.api._v2.keras import initializers
> # pylint: enable=g-import-not-at-top
> ```

It doesn't work for me

**Comment by bschnurr** [comment]
try this
```python
import tensorflow as tf

tf.keras 

```

**Comment by bschnurr** [comment]
could also try latest rc release of tensorflow with these changes. https://github.com/tensorflow/tensorflow/releases/tag/v2.10.0-rc1

**Comment by nyngwang** [comment]
@bschnurr Just confirmed that it works on `tensorflow-gpu v2.10.0-rc1` (installed via `conda`):

<img src="https://user-images.githubusercontent.com/24765272/185237375-3c08f47d-bc4d-442b-b25b-329217f60f92.png" width="400">

**Comment by RafalSkolasinski** [comment]
There seems to still be quite few issues around it. Maybe this should be re-opened until all of them are resolved?

**Comment by bschnurr** [comment]
new stable release has the partial fix. 
https://github.com/tensorflow/tensorflow/releases/tag/v2.10.0

**Comment by nyngwang** [comment]
@bschnurr Could you elaborate more on the partial fix? From the link, I still don't know what has been fixed.

**Comment by bingoct** [comment]
my enviroment use tensorflow-directml == 1.15.8. 
How cloud i fix it?
![image](https://user-images.githubusercontent.com/47520890/195578090-c9599514-0f3d-4895-ab5d-ff790be2361a.png)

**Comment by bschnurr** [comment]
just need to change how you access keras.

grab it from `tf` 

```python
import tensorflow as tf
keras = tf.keras
```

**Comment by jmp75** [comment]
Thank you @bschnurr , this seems to work for me.
 
Pylance confused by:

```python
from tensorflow.keras.metrics import import mean_absolute_error, mean_absolute_percentage_error
```
but fine with:
```python
import tensorflow as tf
keras = tf.keras
from keras.metrics import mean_absolute_error, mean_absolute_percentage_error
```

Go figure.

## Environment

* Linux Debian
* Conda env, but with tensoflow installed from pip (as per tensorflow install instructions)

```
cudatoolkit               11.2.2              hbe64b41_10    conda-forge
cudnn                     8.1.0.77             h90431f1_0    conda-forge
tensorboard               2.10.1                   pypi_0    pypi
tensorboard-data-server   0.6.1                    pypi_0    pypi
tensorboard-plugin-wit    1.8.1                    pypi_0    pypi
tensorflow                2.10.0                   pypi_0    pypi
tensorflow-estimator      2.10.0                   pypi_0    pypi
tensorflow-io-gcs-filesystem 0.27.0                   pypi_0    pypi
```

* vscode:

```
Version: 1.73.0-insider
Commit: 266ff5b78fc4a128994194e828e55b2bbeb0acbe
Date: 2022-10-20T05:20:13.600Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Linux x64 5.10.0-19-amd64
Sandboxed: Yes
```

* pylance:  v2022.10.30  (note: which is not the pre-release version)

**Comment by bschnurr** [comment]
tensorflow has some dynamic behaviour so we can only statically determine the keras module from tf.keras

**Comment by derrick-daniel** [comment]
```python
from keras.metrics import mean_absolute_error, mean_absolute_percentage_error
```

This line is potentially problematic. When you use from keras.metrics import ..., Python might attempt to import from the standalone Keras package (if it's installed) instead of TensorFlow's Keras (tf.keras). This could lead to compatibility issues or unexpected behavior, especially if the standalone Keras and TensorFlow versions are different.

**Comment by nomaan-2k** [comment]
@bschnurr , I don't think `keras = tf.keras` is doing anything here

when we do `from keras import layers`

the `keras` alias is unused here

tried changing it to something random like this

```python
import tensorflow as tf
kfdf = tf.keras 
from keras.preprocessing import image
``` 

still works


Is there any permanent fix to this ?? (except for workarounds)

This issue also exists in google colab (so I don't think it has got much to do with PyLance)
<img width="479" alt="image" src="https://github.com/microsoft/pylance-release/assets/52543604/a1344c45-44e2-4ca7-b83d-dbb2a282e36b">

**Comment by thomsm17** [comment]
@nomaan-2k 
I found a solution from @romerojhh: "
To everyone that's still couldn't make it work. Try this, I just tweaked the popular solution on this thread:

modify the file site-packages/tensorflow/__init__.py near line 387
From:

_keras_module = "keras.api._v2.keras"
keras = _LazyLoader("keras", globals(), _keras_module)
_module_dir = _module_util.get_parent_dir_for_name(_keras_module)
if _module_dir:
  _current_module.__path__ = [_module_dir] + _current_module.__path__
setattr(_current_module, "keras", keras)
To:

Note that I used _keras instead of keras

import typing as _typing
if _typing.TYPE_CHECKING:
  from keras.api._v2 import keras
else:
  _keras_module = "keras.api._v2.keras"
  _keras = _LazyLoader("keras", globals(), _keras_module)
  _module_dir = _module_util.get_parent_dir_for_name(_keras_module)
  if _module_dir:
    _current_module.__path__ = [_module_dir] + _current_module.__path__
  setattr(_current_module, "keras", _keras)
And then make sure to copy folder
/site-packages/keras/ into /site-packages/tensorflow/

After all of that, invalidate and restart your IDE

This works on my setup using Conda"

It worked for me using: Tensorflow 2.13.0 in VS code
Seems to be a permanent fix.

**Comment by mezotaken** [comment]
The only thing that worked for me in the end is this:
```
from keras._tf_keras.keras import *
...
inner = layers.Activation(self.activation, name='act1')(inner)
inner = layers.MaxPooling2D(pool_size = (self.pool_size, self.pool_size), name='max1')(inner)
``` 
This may be temporary or wrong from the perspective of tensorflow devs, yet it works with `tensorflow==2.17.0` and its keras.
Pylance integrated in VSCode.

**Comment by prashantkumsrivastava** [comment]
this is how it worked for me in both **Tensorflow = 2.17** and **keras-3**, if directly installing 
seems you have to import whole class and then  access using fully qualified names 

```
from keras._tf_keras.keras import models ,layers ,optimizers 
model = models.Sequential()
layers.Dense()
```

**Comment by Spider-netizen** [comment]
I also experience issues where pylance doesn't seem to recognize `tf.keras`. Why hasn't this issue been reopened?

**Comment by rchiodo** [comment]
This is happening because tensorflow is dynamically importing keras. Pylance can't follow dynamic imports (imports based on running code), so you need to [workaround](https://github.com/microsoft/pylance-release/issues/1066#issuecomment-2266739431) it or ask tensorflow to change how keras is imported.
